File: clients/client-go/conflation_wire_test.go
package queen

import (
	"context"
	"net/url"
	"strings"
	"sync"
	"testing"
	"time"
)

// Conflation, client side (PLAN_CONFLATION.md §4, the client-go row; §7.2 "the
// same three tests per SDK").
//
// The three things a client can be wrong about, and why each one is asserted
// against the raw query string rather than against "it worked":
//
//  1. THE FLAG MUST REACH THE WIRE FROM BOTH BUILDERS. pop() and consume() have
//     SEPARATE param builders in this SDK (QueueBuilder.buildPopParams vs
//     ConsumerManager.buildParams) — the hazard the plan calls out in §4 and the
//     one a JS bug of exactly this shape already left a comment about. A test
//     that only drives consume() would pass with a pop() that silently drops the
//     option.
//
//  2. DEGRADING MUST BE LOUD. Against a broker older than 1.1.0 the unknown
//     query param is ignored, the pop returns the WHOLE backlog, and a consumer
//     that asked for last-value delivery quietly processes every stale message.
//     Nothing in the response, the status code or the message shape says so. The
//     only usable signal is the ECHO — the broker emits "conflation":true on
//     every conflating pop, including empty ones — so its absence is the error
//     (§4 blockquote).
//
//  3. A DECLARATION CONFLICT MUST NOT BE THE SAME EVENT AS AN OLD BROKER.
//     conflict = a NEW broker answering "the group is already registered the
//     other way, my stored value wins" (§3.3). It is warned about once and the
//     consumer keeps working (§7.3 E2E-4: "both consumers keep working"), which
//     is what makes rolling deploys survivable. Only the absence of BOTH keys is
//     an old broker.

func conflationCtx(t *testing.T) context.Context {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	t.Cleanup(cancel)
	return ctx
}

// queryOf parses one captured request's query string.
func queryOf(t *testing.T, req capturedRequest) url.Values {
	t.Helper()
	v, err := url.ParseQuery(req.RawQuery)
	if err != nil {
		t.Fatalf("captured query is not parseable: %v (%q)", err, req.RawQuery)
	}
	return v
}

// popRequests returns the captured requests that are pops, so a consume test can
// ignore the acks the auto-ack loop interleaves.
func popRequests(reqs []capturedRequest) []capturedRequest {
	var out []capturedRequest
	for _, r := range reqs {
		if strings.HasPrefix(r.Path, "/api/v1/pop") {
			out = append(out, r)
		}
	}
	return out
}

// captureConflationWarnings swaps the once-per-(queue,group) warning sink for a
// counter. The registry that gates it is process-global by design (§4: "one
// warning per (queue,group) per process"), so every test below uses its own
// queue name instead of resetting it — which also proves the key is the pair and
// not a single global latch.
func captureConflationWarnings(t *testing.T) func() []string {
	t.Helper()
	var mu sync.Mutex
	var seen []string
	prev := conflationConflictWarn
	conflationConflictWarn = func(queue, group string) {
		mu.Lock()
		defer mu.Unlock()
		seen = append(seen, queue+"/"+group)
	}
	t.Cleanup(func() { conflationConflictWarn = prev })
	return func() []string {
		mu.Lock()
		defer mu.Unlock()
		out := make([]string, len(seen))
		copy(out, seen)
		return out
	}
}

// --- 1. the flag reaches the wire, from both builders ----------------------

func TestConflationReachesTheWireFromPop(t *testing.T) {
	srv := newCaptureServer(t, okJSON(`{"messages":[],"conflation":true}`))
	client := newWireClient(t, srv.URL)

	if _, err := client.Queue("cfl-pop").Group("workers").Conflation(true).Wait(false).Pop(conflationCtx(t)); err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := queryOf(t, srv.only(t))
	if got := q.Get("conflation"); got != "true" {
		t.Fatalf("pop query conflation = %q, want \"true\" (raw: %q)", got, srv.only(t).RawQuery)
	}
}

func TestConflationIsAbsentFromThePopWireWhenNotRequested(t *testing.T) {
	// Byte-identical requests for every consumer that does not opt in (§8,
	// "default off"): the param is emitted ONLY when true, never as
	// conflation=false.
	srv := newCaptureServer(t, okJSON(`{"messages":[]}`))
	client := newWireClient(t, srv.URL)

	if _, err := client.Queue("cfl-pop-off").Group("workers").Conflation(false).Wait(false).Pop(conflationCtx(t)); err != nil {
		t.Fatalf("pop: %v", err)
	}

	q := queryOf(t, srv.only(t))
	if _, present := q["conflation"]; present {
		t.Fatalf("conflation must not appear on the wire when off; raw query: %q", srv.only(t).RawQuery)
	}
}

func TestConflationReachesTheWireFromConsume(t *testing.T) {
	// Drives the real consume loop, not buildParams directly: the point is to
	// prove the whole path (builder field -> ConsumeOptions -> buildParams).
	srv := newCaptureServer(t, okJSON(`{"messages":[{"transactionId":"11111111-1111-4111-8111-111111111111","partitionId":"22222222-2222-4222-8222-222222222222","queue":"cfl-consume","partition":"Default","data":{"n":1},"createdAt":"2026-08-21T10:00:00.000Z"}],"conflation":true}`))
	client := newWireClient(t, srv.URL)

	var handled int
	err := client.Queue("cfl-consume").
		Group("workers").
		Conflation(true).
		Wait(false).
		Limit(1).
		Consume(conflationCtx(t), func(ctx context.Context, msg *Message) error {
			handled++
			return nil
		}).Execute(conflationCtx(t))
	if err != nil {
		t.Fatalf("consume: %v", err)
	}
	if handled != 1 {
		t.Fatalf("handler ran %d times, want 1", handled)
	}

	pops := popRequests(srv.requests())
	if len(pops) == 0 {
		t.Fatal("consume made no pop request")
	}
	q := queryOf(t, pops[0])
	if got := q.Get("conflation"); got != "true" {
		t.Fatalf("consume query conflation = %q, want \"true\" (raw: %q)", got, pops[0].RawQuery)
	}
}

func TestConflationIsAbsentFromTheConsumeWireWhenNotRequested(t *testing.T) {
	srv := newCaptureServer(t, okJSON(`{"messages":[]}`))
	client := newWireClient(t, srv.URL)

	params := NewConsumerManager(client.httpClient, client).buildParams(
		client.Queue("cfl-consume-off").Group("workers").getConsumeOptions())
	v, err := url.ParseQuery(params)
	if err != nil {
		t.Fatalf("params are not parseable: %v (%q)", err, params)
	}
	if _, present := v["conflation"]; present {
		t.Fatalf("conflation must not appear on the wire when off; params: %q", params)
	}
}

// --- 2. degrade loudly ------------------------------------------------------

func TestPopErrorsWhenTheBrokerDoesNotEchoConflation(t *testing.T) {
	// The old-broker case: the param was ignored, so neither key comes back.
	// This must be an ERROR and not a warning — a warning on stderr is exactly
	// what nobody reads while the consumer chews through a 4M backlog.
	srv := newCaptureServer(t, okJSON(`{"messages":[]}`))
	client := newWireClient(t, srv.URL)

	_, err := client.Queue("cfl-old-pop").Group("workers").Conflation(true).Wait(false).Pop(conflationCtx(t))
	if err == nil {
		t.Fatal("pop against a broker that ignored conflation returned no error")
	}
	if !strings.Contains(err.Error(), "requires broker >= 1.1.0") {
		t.Fatalf("error does not name the version requirement: %v", err)
	}
}

func TestPopDoesNotErrorWhenConflationWasNotRequested(t *testing.T) {
	srv := newCaptureServer(t, okJSON(`{"messages":[]}`))
	client := newWireClient(t, srv.URL)

	if _, err := client.Queue("cfl-noflag").Group("workers").Wait(false).Pop(conflationCtx(t)); err != nil {
		t.Fatalf("a pop that never asked for conflation must not care about the echo: %v", err)
	}
}

func TestConsumeStopsWhenTheBrokerDoesNotEchoConflation(t *testing.T) {
	// The echo is emitted on EMPTY pops too, so the loop dies on the first
	// round trip — before a single stale message is handed to the handler.
	srv := newCaptureServer(t,
		okJSON(`{"messages":[{"transactionId":"11111111-1111-4111-8111-111111111111","partitionId":"22222222-2222-4222-8222-222222222222","queue":"cfl-old-consume","partition":"Default","data":{"n":1},"createdAt":"2026-08-21T10:00:00.000Z"}]}`),
	)
	client := newWireClient(t, srv.URL)

	var handled int
	err := client.Queue("cfl-old-consume").
		Group("workers").
		Conflation(true).
		Wait(false).
		Consume(conflationCtx(t), func(ctx context.Context, msg *Message) error {
			handled++
			return nil
		}).Execute(conflationCtx(t))

	if err == nil {
		t.Fatal("consume against a broker that ignored conflation ran on without an error")
	}
	if !strings.Contains(err.Error(), "requires broker >= 1.1.0") {
		t.Fatalf("error does not name the version requirement: %v", err)
	}
	if handled != 0 {
		t.Fatalf("handler ran %d times; a degraded pop must not be processed", handled)
	}
	if n := len(popRequests(srv.requests())); n != 1 {
		t.Fatalf("consume made %d pops, want exactly 1 (the loop must stop on the first)", n)
	}
}

// --- 3. the conflict warning: once per (queue, group), and NOT an error ------

func TestConflationConflictWarnsExactlyOncePerQueueAndGroup(t *testing.T) {
	warnings := captureConflationWarnings(t)
	srv := newCaptureServer(t,
		okJSON(`{"messages":[],"conflation":true,"conflationConflict":true}`),
		okJSON(`{"messages":[],"conflation":true,"conflationConflict":true}`),
		okJSON(`{"messages":[],"conflation":true,"conflationConflict":true}`),
	)
	client := newWireClient(t, srv.URL)
	ctx := conflationCtx(t)

	for i := 0; i < 3; i++ {
		if _, err := client.Queue("cfl-conflict").Group("workers").Conflation(true).Wait(false).Pop(ctx); err != nil {
			t.Fatalf("pop %d: %v", i, err)
		}
	}

	got := warnings()
	if len(got) != 1 {
		t.Fatalf("emitted %d warnings %v, want exactly 1", len(got), got)
	}
	if got[0] != "cfl-conflict/workers" {
		t.Fatalf("warning keyed on %q, want \"cfl-conflict/workers\"", got[0])
	}
}

func TestConflationConflictWarnsAgainForADifferentGroup(t *testing.T) {
	warnings := captureConflationWarnings(t)
	srv := newCaptureServer(t,
		okJSON(`{"messages":[],"conflation":true,"conflationConflict":true}`),
		okJSON(`{"messages":[],"conflation":true,"conflationConflict":true}`),
	)
	client := newWireClient(t, srv.URL)
	ctx := conflationCtx(t)

	for _, group := range []string{"alpha", "beta"} {
		if _, err := client.Queue("cfl-conflict-groups").Group(group).Conflation(true).Wait(false).Pop(ctx); err != nil {
			t.Fatalf("pop %s: %v", group, err)
		}
	}

	got := warnings()
	if len(got) != 2 {
		t.Fatalf("emitted %d warnings %v, want 2 (the gate is per (queue,group), not global)", len(got), got)
	}
}

func TestConflationConflictIsNotAnOldBroker(t *testing.T) {
	// A conflict where the STORED value is false: the broker answers with
	// conflationConflict but no "conflation":true, because the effective flag is
	// false. That is a live 1.1.0 broker telling this consumer the group already
	// exists the other way — group-setting-wins, warn, keep working (§3.3, Q3).
	// Treating it as an old broker would take down exactly the half of a rolling
	// fleet that is already correct.
	warnings := captureConflationWarnings(t)
	srv := newCaptureServer(t, okJSON(`{"messages":[],"conflationConflict":true}`))
	client := newWireClient(t, srv.URL)

	if _, err := client.Queue("cfl-conflict-stored-off").Group("workers").Conflation(true).Wait(false).Pop(conflationCtx(t)); err != nil {
		t.Fatalf("a declaration conflict must not fail the pop: %v", err)
	}
	if got := warnings(); len(got) != 1 {
		t.Fatalf("emitted %d warnings %v, want exactly 1", len(got), got)
	}
}

// --- depth: the two-number surface reaches the caller -----------------------

func TestQueueDepthCarriesTheConflationFields(t *testing.T) {
	// GetQueueDepth hands back the decoded body, so the three fields of §5.3
	// ride through for free — which is exactly why this is worth pinning: the
	// day someone types this response into a struct, the fields that make a
	// 4M-deep conflating queue readable as HEALTHY would vanish silently.
	srv := newCaptureServer(t, okJSON(`{"queue":"cfl-depth","group":"workers","pending":4000000,"partitionsPending":12,"conflation":true,"effectivePending":12,"partitions":[{"partition":"Default","pending":4000000}]}`))
	client := newWireClient(t, srv.URL)

	depth, err := client.Admin().GetQueueDepth(conflationCtx(t), "cfl-depth", "workers")
	if err != nil {
		t.Fatalf("depth: %v", err)
	}
	if got := srv.only(t).Path; got != "/api/v1/resources/queues/cfl-depth/depth" {
		t.Fatalf("depth went to %q", got)
	}
	if conflating, _ := depth["conflation"].(bool); !conflating {
		t.Fatalf("conflation = %v, want true", depth["conflation"])
	}
	for _, field := range []string{"partitionsPending", "effectivePending"} {
		v, ok := depth[field].(float64)
		if !ok {
			t.Fatalf("%s missing from the depth payload (%v)", field, depth[field])
		}
		if v != 12 {
			t.Fatalf("%s = %v, want 12", field, v)
		}
	}
	// The distinction is the whole point: log depth stays the big number.
	if pending, _ := depth["pending"].(float64); pending != 4000000 {
		t.Fatalf("pending = %v, want 4000000 (log depth, not work depth)", depth["pending"])
	}
}

// --- streams: the option survives the Source adapter ------------------------

func TestConflationThreadsThroughTheStreamSourceAdapter(t *testing.T) {
	srv := newCaptureServer(t, okJSON(`{"messages":[],"conflation":true}`))
	client := newWireClient(t, srv.URL)

	src := client.Queue("cfl-stream").Group("workers").Wait(false).AsStreamSource().Conflation(true)
	if _, err := src.Pop(conflationCtx(t)); err != nil {
		t.Fatalf("stream source pop: %v", err)
	}

	q := queryOf(t, srv.only(t))
	if got := q.Get("conflation"); got != "true" {
		t.Fatalf("stream source query conflation = %q, want \"true\" (raw: %q)", got, srv.only(t).RawQuery)
	}
}

File: clients/client-js/test-v2/conflation-unit/conflationWire.test.js
/**
 * Conflation: wire contract + degrade-loudly (PLAN_CONFLATION.md §3.1, §3.3, §4).
 *
 * Three things are pinned here, and each of them is a defect that ships
 * silently if it is not:
 *
 *   1. `conflation` reaches the wire from BOTH entry points. pop() and
 *      consume() build their query strings in two SEPARATE places
 *      (QueueBuilder.pop's inline params and ConsumerManager#buildParams) —
 *      the standing comment at QueueBuilder.js documents a bug of exactly this
 *      shape, where the parameter was added to one builder and the other kept
 *      working while the option "did nothing". Both directions are asserted,
 *      including the negative: an undeclared conflation must not add the key at
 *      all, because the broker's response bytes are pinned byte-for-byte for
 *      flag-off deployments.
 *
 *   2. Degrade-loudly (§4). No SDK negotiates a version, so a new client against
 *      a pre-1.1.0 broker sends `conflation=true`, the broker ignores the unknown
 *      query parameter, and the consumer quietly drains the whole backlog one
 *      message at a time. The detection is response-driven: a conflating broker
 *      echoes `"conflation":true` on EVERY response, empty ones included (it
 *      answers 200-with-body instead of 204 precisely so the echo has somewhere
 *      to ride), so the first round trip is enough. An old broker's empty pop is
 *      a bodiless 204 => `null` here, which is why null must raise too.
 *
 *   3. Declaration conflict (§3.3). The STORED group policy wins; the consumer
 *      that disagreed keeps working and warns exactly ONCE per (queue, group)
 *      per process. Rejecting instead would take down the correct half of a
 *      rolling deploy (plan Q3), and warning per response would flood.
 *
 * Same style as http-unit/retry429.test.js and kv-unit/kvWire.test.js: a real
 * node:http server playing a canned plan, real fetch, real JSON. No mocking
 * framework, and every assertion is about bytes that actually crossed a socket.
 */

import { describe, it, beforeEach, afterEach } from 'node:test'
import assert from 'node:assert/strict'

import { Queen } from '../../client-v2/index.js'
import { Runner } from '../../client-v2/streams/runtime/Runner.js'
import * as logger from '../../client-v2/utils/logger.js'
import {
  CONFLATION_UNSUPPORTED,
  resetConflationWarnings
} from '../../client-v2/utils/conflation.js'
import { withPlanServer, ok } from '../kv-unit/_planServer.js'

const QUEUE = 'test-conflation'
const GROUP = 'workers'

/** One delivered frame, shaped like a real pop response element. */
function frame(n = 1) {
  return {
    transactionId: `txn-${n}`,
    partitionId: `part-${n}`,
    partition: 'Default',
    payload: { n },
    leaseId: 'lease-1',
    consumerGroup: GROUP
  }
}

/** A 200 pop body; `extra` carries the conflation echo keys under test. */
function popBody(extra = {}, frames = [frame()]) {
  return ok({ messages: frames, partitionsClaimed: frames.length, ...extra })
}

/** A pre-1.1.0 broker's empty pop: bodiless 204 (HttpClient turns it into null). */
const emptyNoContent = { status: 204 }

/** Query parameters of a recorded hit. */
function query(url) {
  const i = url.indexOf('?')
  return new URLSearchParams(i < 0 ? '' : url.slice(i + 1))
}

/** Path (without query) of a recorded hit. */
function path(url) {
  const i = url.indexOf('?')
  return i < 0 ? url : url.slice(0, i)
}

async function withQueen(plan, defaultResponse, run) {
  await withPlanServer(plan, defaultResponse, async (url, hits) => {
    const queen = new Queen({ url, handleSignals: false })
    try {
      await run(queen, hits)
    } finally {
      await queen.close()
    }
  })
}

// The conflict warning is deliberately once-per-process state, so every test
// starts from a clean registry.
beforeEach(() => { resetConflationWarnings() })
afterEach(() => { resetConflationWarnings(); logger.configure(null) })

// ---------------------------------------------------------------------------
// 1. The option reaches the wire — from pop() AND from consume().
// ---------------------------------------------------------------------------

describe('conflation — wire (pop)', () => {
  it('pop() sends conflation=true when the builder declares it', async () => {
    await withQueen([popBody({ conflation: true })], popBody({ conflation: true }), async (queen, hits) => {
      const messages = await queen.queue(QUEUE).group(GROUP).conflation(true).pop()

      assert.equal(messages.length, 1)
      assert.equal(hits.length, 1)
      assert.equal(path(hits[0].url), `/api/v1/pop/queue/${QUEUE}`)
      assert.equal(query(hits[0].url).get('conflation'), 'true')
      assert.equal(query(hits[0].url).get('consumerGroup'), GROUP)
    })
  })

  it('pop() omits the parameter entirely when conflation is not declared', async () => {
    await withQueen([popBody()], popBody(), async (queen, hits) => {
      await queen.queue(QUEUE).group(GROUP).pop()

      assert.equal(hits.length, 1)
      assert.equal(
        query(hits[0].url).has('conflation'),
        false,
        'a flag-off pop must be byte-identical to today: no conflation key at all'
      )
    })
  })
})

describe('conflation — wire (consume)', () => {
  it('consume() sends conflation=true through ConsumerManager#buildParams', async () => {
    await withQueen([popBody({ conflation: true })], popBody({ conflation: true }), async (queen, hits) => {
      let handled = 0
      await queen.queue(QUEUE)
        .group(GROUP)
        .conflation(true)
        .wait(false)
        .autoAck(false)
        .limit(1)
        .consume(async () => { handled++ })

      assert.equal(handled, 1)
      assert.ok(hits.length >= 1)
      assert.equal(path(hits[0].url), `/api/v1/pop/queue/${QUEUE}`)
      assert.equal(
        query(hits[0].url).get('conflation'),
        'true',
        'consume() builds its params in a DIFFERENT place than pop() — both must carry it'
      )
    })
  })

  it('consume() omits the parameter entirely when conflation is not declared', async () => {
    await withQueen([popBody()], popBody(), async (queen, hits) => {
      await queen.queue(QUEUE)
        .group(GROUP)
        .wait(false)
        .autoAck(false)
        .limit(1)
        .consume(async () => {})

      assert.ok(hits.length >= 1)
      assert.equal(query(hits[0].url).has('conflation'), false)
    })
  })
})

// ---------------------------------------------------------------------------
// 2. Degrade-loudly (§4): an old broker must never be silent.
// ---------------------------------------------------------------------------

describe('conflation — degrade loudly against a pre-1.1.0 broker', () => {
  it('pop() raises when the response carries messages but no conflation echo', async () => {
    await withQueen([popBody()], popBody(), async (queen) => {
      await assert.rejects(
        () => queen.queue(QUEUE).group(GROUP).conflation(true).pop(),
        (err) => {
          assert.equal(err.code, CONFLATION_UNSUPPORTED)
          assert.match(err.message, /requires broker >= 1\.1\.0/)
          return true
        }
      )
    })
  })

  it('pop() raises on the FIRST empty pop, which an old broker answers 204', async () => {
    await withQueen([emptyNoContent], emptyNoContent, async (queen, hits) => {
      await assert.rejects(
        () => queen.queue(QUEUE).group(GROUP).conflation(true).pop(),
        (err) => {
          assert.equal(err.code, CONFLATION_UNSUPPORTED)
          return true
        }
      )
      assert.equal(hits.length, 1, 'the error fires on the first round trip, before any message')
    })
  })

  it('consume() stops the loop and rejects before the handler ever runs', async () => {
    await withQueen([popBody()], popBody(), async (queen, hits) => {
      let handled = 0
      // consume() returns a thenable ConsumeBuilder, not a Promise: await it
      // inside the arrow so assert.rejects sees a real promise.
      await assert.rejects(
        async () => {
          await queen.queue(QUEUE)
            .group(GROUP)
            .conflation(true)
            .wait(false)
            .autoAck(false)
            .consume(async () => { handled++ })
        },
        (err) => {
          assert.equal(err.code, CONFLATION_UNSUPPORTED)
          assert.match(err.message, /requires broker >= 1\.1\.0/)
          return true
        }
      )
      assert.equal(handled, 0, 'not one message may be processed under a silent degrade')
      assert.equal(hits.length, 1, 'the loop stops instead of draining the backlog')
    })
  })

  it('leaves a conflation-less response alone when conflation was never requested', async () => {
    await withQueen([popBody()], popBody(), async (queen) => {
      const messages = await queen.queue(QUEUE).group(GROUP).pop()
      assert.equal(messages.length, 1, 'today\'s behaviour is untouched when the flag is off')
    })
  })
})

// ---------------------------------------------------------------------------
// 3. Declaration conflict (§3.3): the stored policy wins, loudly but once.
// ---------------------------------------------------------------------------

/** Collect every warning the SDK emits while `run` executes. */
async function captureWarnings(run) {
  const warnings = []
  logger.configure({
    info() {},
    warn(entry) { warnings.push(typeof entry === 'string' ? entry : JSON.stringify(entry)) },
    error() {}
  })
  try {
    await run()
  } finally {
    logger.configure(null)
  }
  return warnings
}

describe('conflation — declaration conflict', () => {
  it('warns exactly once per (queue, group) and keeps delivering messages', async () => {
    const conflicted = popBody({ conflationConflict: true })
    await withQueen([conflicted, conflicted, conflicted], conflicted, async (queen, hits) => {
      const warnings = await captureWarnings(async () => {
        for (let i = 0; i < 3; i++) {
          const messages = await queen.queue(QUEUE).group(GROUP).conflation(true).pop()
          assert.equal(messages.length, 1, 'a conflict never costs the consumer its messages')
        }
      })

      assert.equal(hits.length, 3)
      const conflictWarnings = warnings.filter(w => w.includes('Conflation.conflict'))
      assert.equal(
        conflictWarnings.length,
        1,
        `exactly one warning per (queue,group) per process, got: ${JSON.stringify(conflictWarnings)}`
      )
      assert.match(conflictWarnings[0], new RegExp(QUEUE))
      assert.match(conflictWarnings[0], new RegExp(GROUP))
    })
  })

  it('warns again for a different group on the same queue', async () => {
    const conflicted = popBody({ conflationConflict: true })
    await withQueen([conflicted, conflicted], conflicted, async (queen) => {
      const warnings = await captureWarnings(async () => {
        await queen.queue(QUEUE).group(GROUP).conflation(true).pop()
        await queen.queue(QUEUE).group('audit').conflation(true).pop()
      })

      const conflictWarnings = warnings.filter(w => w.includes('Conflation.conflict'))
      assert.equal(conflictWarnings.length, 2, 'the warn-once key is (queue, group), not queue')
    })
  })

  it('a conflict is never a degrade error — the broker plainly understood the flag', async () => {
    const conflicted = popBody({ conflationConflict: true })
    await withQueen([conflicted], conflicted, async (queen) => {
      const messages = await queen.queue(QUEUE).group(GROUP).conflation(true).pop()
      assert.equal(messages.length, 1)
    })
  })
})

// ---------------------------------------------------------------------------
// 4. The streams runtime pops through the same QueueBuilder — so the option has
//    to be threaded there too, or `run({ conflation: true })` is a no-op.
// ---------------------------------------------------------------------------

/** A QueueBuilder-shaped source that records the chain the Runner builds. */
function recordingSource(calls) {
  const qb = {
    batch(v) { calls.push(['batch', v]); return qb },
    wait(v) { calls.push(['wait', v]); return qb },
    timeoutMillis(v) { calls.push(['timeoutMillis', v]); return qb },
    group(v) { calls.push(['group', v]); return qb },
    partitions(v) { calls.push(['partitions', v]); return qb },
    subscriptionMode(v) { calls.push(['subscriptionMode', v]); return qb },
    subscriptionFrom(v) { calls.push(['subscriptionFrom', v]); return qb },
    conflation(v) { calls.push(['conflation', v]); return qb },
    async pop() { calls.push(['pop']); return [] }
  }
  return qb
}

function runnerWith(opts, calls) {
  return new Runner({
    url: 'http://127.0.0.1:1',
    queryId: 'test-conflation-query',
    stream: { source: recordingSource(calls), stages: {}, config_hash: 'hash' },
    ...opts
  })
}

describe('conflation — streams runtime', () => {
  it('Stream.run({ conflation: true }) declares it on the source pop', async () => {
    const calls = []
    await runnerWith({ conflation: true }, calls)._popMessages()
    assert.deepEqual(calls.filter(c => c[0] === 'conflation'), [['conflation', true]])
  })

  it('leaves the source pop untouched when the option is absent', async () => {
    const calls = []
    await runnerWith({}, calls)._popMessages()
    assert.deepEqual(calls.filter(c => c[0] === 'conflation'), [])
  })
})

// ---------------------------------------------------------------------------
// 5. Depth: the new fields must survive the admin call (§2.5/§5.3).
// ---------------------------------------------------------------------------

describe('conflation — queue depth', () => {
  it('getQueueDepth surfaces partitionsPending / conflation / effectivePending', async () => {
    const depth = ok({
      queue: QUEUE,
      group: GROUP,
      pending: 4000000,
      partitionsPending: 12,
      conflation: true,
      effectivePending: 12,
      partitions: [{ partition: 'a', pending: 333333 }]
    })
    await withQueen([depth], depth, async (queen, hits) => {
      const got = await queen.admin.getQueueDepth(QUEUE, GROUP)

      assert.equal(path(hits[0].url), `/api/v1/resources/queues/${QUEUE}/depth`)
      assert.equal(query(hits[0].url).get('group'), GROUP)
      // Log depth vs work depth: 4M positions to retire, 12 handler calls owed.
      assert.equal(got.pending, 4000000)
      assert.equal(got.partitionsPending, 12)
      assert.equal(got.conflation, true)
      assert.equal(got.effectivePending, 12)
    })
  })
})
